feat: c at_activate and required components

srieteja opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines in CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->
Closes https://github.com/atsign-foundation/at_c/issues/176 and https://github.com/atsign-foundation/noports/issues/1325. Partially satisfies https://github.com/atsign-foundation/at_c/issues/399

**- What I did**
- ATAUTH introduction

1. Introduced implementation for atactivate.
2. Introduced package atauth that holds impl for atactivate, atactivate-arg-parser, sending enroll request and fetching home dir.

- ATCHOPS changes

1. Introduced hex_utils in atchops that has impl to convert data from byte-to-hex and hex-to-bytes.
2. Allow SHA512 as a valid hash type in atchops_sha_hash().
3. Introduced utf8 utils in atchops that utf8 encodes string data (converts string to bytes).

- ATCLIENT changes

1. Introduce impl to perform cram authentication.
2. Introduced method to create/construct reserved keys.
3. Rename atclient_pkam_authenticate_options -> atclient_authenticate_options.
4. Rename ATCLIENT_PKAM_AUTHENTICATE_DEFAULT_ROOT_SERVER_HOST -> ATCLIENT_AUTHENTICATE_DEFAULT_ROOT_SERVER_HOST
5. Rename ATCLIENT_PKAM_AUTHENTICATE_DEFAULT_ROOT_SERVER_PORT -> ATCLIENT_AUTHENTICATE_DEFAULT_ROOT_SERVER_PORT
6. Introduced new param in atclient_delete_request_options named "skip_shared_by_check" and supporting method: init, set, unset, free and is_initialized.
7. Updated atclient_delete impl to ensure that keys can be used without "shared_by" param. 
8. Updated atkey.c to ensure that is_shared_by_initialized check is skipped for reserved keys.
9. Added impl in connection.c to ensure that when mbedtls_ssl_write performs a partial write, the rest of the data is recursively written until the entire data is written.
10. Added error_buffers and pass them to mbedtls_ssl_read and mbedtls_ssl_write to get accurate error messages in case of failure

- ATCOMMONS introduction
1. Introduced package atcommons that currently hold impl for the following: 

> enroll_command_builder - constructs an enroll:request command based on the params passed
> enroll_namespace_struct - struct to hold enroll namespace  and enroll_namespace_list. Also has a toJson method.
> enroll_params_struct - struct to hold enroll params. Has init and toJson methods
> enroll_operation_enum - enum to specify enroll operation. Has a toString method
> enroll_status_enum - enum to specify enrollment status. Has a toString method

**- How to verify it**
Tests on the way

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Implement atactivate, enroll:request, cram auth support and other minor components required by atactivate

srieteja committed:
feat: create atcommons and introduce enroll related deps

xavierchanth committed:
feat: add apkam_symmetric_key to atkeys_file

xavierchanth committed:
feat: add apkam_symmetric_key to atkeys

srieteja committed:
feat: introduce arg parser for at_activate

srieteja committed:
feat: introduce atactivate + hex, base64 and utf8 utils

srieteja committed:
feat: cmakeLists update + default homedir impl + more atactivate code

srieteja committed:
docs: added docs + some more refactoring

srieteja committed:
fix: migration to mallocs + minor refactoring

srieteja committed:
chore: resolve compile warnings and refactor methods signatures

srieteja committed:
fix: import platform dependent header limits.

srieteja committed:
chore: rename all instance of "atclient_pkam_authenticate_options_*" -> "atclient_authenticate_options_*"

dependabot[bot] committed:
chore(deps): bump github/codeql-action in the github-actions group

Bumps the github-actions group with 1 update: [github/codeql-action](https://github.com/github/codeql-action).


Updates `github/codeql-action` from 3.27.1 to 3.27.3
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/4f3212b61783c3c68e8309a0f18a699764811cda...396bb3e45325a47dd9ef434068033c6d5bb0d11a)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: github-actions
...

Signed-off-by: dependabot[bot] <support@github.com>

srieteja committed:
fix: analyzer warnings

srieteja committed:
chore: update .gitignore

srieteja committed:
chore: fix method duplication

srieteja commented on packages/atclient/src/atkey.c:
```diff
@@ -73,19 +73,22 @@ size_t atclient_atkey_strlen(const atclient_atkey *atkey) {
     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atkey is NULL\n");
     return 0;
   }
```
Separate if-clauses that check if 'key' and 'shared_by' are initialized. Also, accept if reserved keys do not have shared_by initialized

srieteja committed:
chore: refactoring + add missing docs + better logging

dependabot[bot] committed:
chore(deps): bump github/codeql-action in the github-actions group

Bumps the github-actions group with 1 update: [github/codeql-action](https://github.com/github/codeql-action).


Updates `github/codeql-action` from 3.27.3 to 3.27.4
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/396bb3e45325a47dd9ef434068033c6d5bb0d11a...ea9e4e37992a54ee68a9622e985e60c8e8f12d9f)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: github-actions
...

Signed-off-by: dependabot[bot] <support@github.com>

srieteja committed:
test: unit tests + refactoring

srieteja committed:
test: unit tests + refactoring

srieteja committed:
Merge remote-tracking branch 'origin/create_at_commons' into create_at_commons

# Conflicts:
#	packages/atcommons/CMakeLists.txt

srieteja commented on packages/atauth/src/atactivate_arg_parser.c:
```diff
@@ -0,0 +1,86 @@
+    default:
+      fprintf(stderr, "Usage: %s -a atsign -c cram-secret [-r root-server] [-p port]\n", argv[0]);
+      ret = -1;
+      goto exit;
```
-h flag

srieteja replied:
change made as part of commit: [ac675cf](https://github.com/atsign-foundation/at_c/pull/423/commits/ac675cf72fe01483c980613bd39e186792bc865d)

srieteja commented on packages/atauth/src/atactivate_arg_parser.c:
```diff
@@ -0,0 +1,86 @@
+      break;
+    case 'p':
+      *root_port = atoi(optarg);
+      break;
```
read otp arg

srieteja replied:
change made as part of commit: [ac675cf](https://github.com/atsign-foundation/at_c/pull/423/commits/ac675cf72fe01483c980613bd39e186792bc865d)

srieteja commented on packages/atclient/include/atclient/constants.h:
```diff
@@ -15,17 +15,21 @@ extern "C" {
 #define DATA_TOKEN "data:"
 
+#define CRAM_PREFIX "cram"
+#define CRAM_COMMAND_LEN 135
```
add ATCLIENT prefix

srieteja replied:
change made as part of commit: [ac675cf](https://github.com/atsign-foundation/at_c/pull/423/commits/ac675cf72fe01483c980613bd39e186792bc865d)

JeremyTubongbanua commented:
Look into args.c in the repl example on how I used `3rdparty/argparse`
https://github.com/atsign-foundation/at_c/blob/trunk/examples/desktop/repl/src/args.c

JeremyTubongbanua commented:
See how I defined a `size` and `len` and sent the `len`

```c
  const size_t update_cmd_size =
      strlen("update") + metadata_protocol_str_len + strlen(":") + atkey_str_len + strlen(" ") + value_encrypted_base64_len + strlen("\r\n") + 1;
  if ((update_cmd = malloc(sizeof(char) * update_cmd_size)) == NULL) {
    ret = 1;
    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to allocate memory for update_cmd\n");
    goto exit;
  }

  snprintf(update_cmd, update_cmd_size, "update%s:%s %s\r\n", metadata_protocol_str, atkey_str, value_encrypted_base64);
  const size_t update_cmd_len = update_cmd_size - 1;

  /*
   * 7. Send update command
   */
  if ((ret = atclient_connection_send(&ctx->atserver_connection, (unsigned char *) update_cmd, update_cmd_len, recv, recv_size,
                                      &recv_len)) != 0) {
    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_connection_send: %d\n", ret);
    goto exit;
  }
```

gkc commented:
@srieteja To clarify requirement a little bit more; the normal sequence for usage of the C at_activate will be as follows:

**1) On some host which is not the device host**
- using the **_dart_** at_activate program
- user onboards their `@device` atSign `at_activate onboard`
  - result: the primary atKeys file is generated
- user generates an enrollment OTP for their device `at_activate otp`
  - result: an OTP is generated and output - e.g. `4A8QFZ`

**2) On the device host**
- using the **_C_** at_activate program
- user runs `at_activate enroll` with parameters the same as those which the dart program uses
  - For example: `at_activate enroll -s 4A8QFZ -p noports -d device_name -n "sshnp:rw,sshrvd:rw" -a @device`
  - sends enroll request and goes into attempt-pkam-auth loop
  - result: an atKeys file unique to this device

So - for our MVP, we need the C at_activate to support `at_activate enroll`

srieteja commented on packages/atclient/src/atclient.c:
```diff
@@ -497,11 +704,39 @@ static int atclient_pkam_authenticate_validate_arguments(const atclient *ctx, co
   }
 
-  if (atsign == NULL || strlen(atsign) == 0) {
+  if (NULL == atsign || strlen(atsign) == 0) {
```
revisit this change

srieteja replied:
change made as part of commit: [ac675cf](https://github.com/atsign-foundation/at_c/pull/423/commits/ac675cf72fe01483c980613bd39e186792bc865d)

xavierchanth replied:
The point of revisiting this change was not to figure out when it was made, but why we got a segfault before it was changed. Masking the problem is not fixing the problem.

srieteja commented on packages/atclient/src/connection.c:
```diff
@@ -402,11 +405,54 @@ int atclient_connection_send(atclient_connection *ctx, const unsigned char *src,
+                 "Partial write successful. Performing recursive write. Current ret: %d\n", ret);
+
+    while (attempt < ATCLIENT_CONNECTION_MAX_RECURSIVE_WRITE_ATTEMPTS) {
+      ret = mbedtls_ssl_write(&(ctx->ssl), src + length_written, remaining_length);
```
do not have a attempt limit

srieteja replied:
change made as part of commit: [ac675cf](https://github.com/atsign-foundation/at_c/pull/423/commits/ac675cf72fe01483c980613bd39e186792bc865d)

srieteja commented on packages/atcommons/include/atcommons/enroll_namespace.h:
```diff
@@ -0,0 +1,64 @@
+} enroll_namespace_t;
+
+typedef struct {
+  int length;
```
use size_t

srieteja replied:
change made as part of commit: [ac675cf](https://github.com/atsign-foundation/at_c/pull/423/commits/ac675cf72fe01483c980613bd39e186792bc865d)

srieteja commented on packages/atcommons/include/atcommons/enroll_namespace.h:
```diff
@@ -0,0 +1,64 @@
+                                       size_t *ns_str_len,
+                                       const enroll_namespace_t *ns);
+
+/**
```
done till here

dependabot[bot] committed:
chore(deps): bump github/codeql-action in the github-actions group

Bumps the github-actions group with 1 update: [github/codeql-action](https://github.com/github/codeql-action).


Updates `github/codeql-action` from 3.27.4 to 3.27.5
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/ea9e4e37992a54ee68a9622e985e60c8e8f12d9f...f09c1c0a94de965c15400f5634aa42fac8fb8f88)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: github-actions
...

Signed-off-by: dependabot[bot] <support@github.com>

srieteja committed:
chore: address review comments

srieteja commented on packages/atcommons/src/command_builders/enroll_command_builder.c:
```diff
@@ -0,0 +1,110 @@
+  }
+  if ((ret = atcommons_enroll_params_to_json(&params_json, params_json_size, &params_json_len, params)) != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atcommons_enroll_params_to_json: %d\n", ret);
+    ret = 1;
```
do not overwrite the ret value

srieteja commented on packages/atcommons/src/enroll/enroll_namespace.c:
```diff
@@ -0,0 +1,74 @@
+
+  if (ns_list_string != NULL) {
+    const char *temp_json_str = cJSON_PrintUnformatted(json_obj);
+    strncpy(ns_list_string, temp_json_str, ns_list_str_size);
```
need not allocate memory for CJSON string manually

srieteja commented on packages/atcommons/include/atcommons/enroll_operation.h:
```diff
@@ -0,0 +1,25 @@
+#define MAX_ENROLL_OPERATION_STRING_LEN 8
+
+typedef enum {
+  apkam_request,
```
use an atcommons_prefix

srieteja commented on packages/atcommons/src/enroll/enroll_operation.c:
```diff
@@ -0,0 +1,48 @@
+
+#define ENROLL_OPERATION_REQUEST "request"
+#define ENROLL_OPERATION_APPROVE "approve"
+#define ENROLL_OPERATION_DENY "deny"
```
atclient prefix

srieteja commented on packages/atcommons/src/enroll/enroll_params.c:
```diff
@@ -0,0 +1,145 @@
+      goto exit;
+    }
+    memset(ns_json, 0, ns_list_str_size);
+    if (ns_json == NULL) {
```
duplicate if-clause

srieteja commented on packages/atcommons/src/enroll/enroll_params.c:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  if (json_string == NULL) {                                     // used to calculate string length
+    *json_string_len = strlen(cJSON_PrintUnformatted(json)) + 1; // +1 for \0
```
cJSON_PrintUnformatted() already allocated mem

srieteja commented on packages/atcommons/src/enroll/enroll_params.c:
```diff
@@ -0,0 +1,145 @@
+                            (const char *)ep->encrypted_default_encryption_private_key);
+  }
+
+  if (ep->encrypted_self_encryption_key) {
```
explicit NULL check

srieteja commented on packages/atcommons/include/atcommons/enroll_status.h:
```diff
@@ -0,0 +1,23 @@
+
+typedef enum {
+  enroll_status_pending,
+  enroll_status_approved,
```
use an atcommons prefix

srieteja commented on packages/atcommons/tests/at_expect.c:
```diff
@@ -0,0 +1,12 @@
+int atcommons_string_expect(char *actual, char*expected){
+  int ret = strcmp(actual, expected);
+  if (ret != 0) {
+    printf("test failed\nexpected: %s\n*actual*: %s\n", expected, actual);
```
replace printf's with atlogger

srieteja commented on packages/atcommons/tests/test_enroll_command_builder.c:
```diff
@@ -0,0 +1,50 @@
+    goto exit;
+  }
+  printf("%s\n", command);
+  printf("%s\n", expected_string);
```
not needed

srieteja commented on packages/atcommons/src/command_builders/enroll_command_builder.c:
```diff
@@ -0,0 +1,110 @@
+#define ENROLL_PREFIX "enroll:"
+#define TAG "enroll command builder"
+
+int atcommons_build_enroll_command(char *command, const size_t cmd_size, size_t *cmd_len,
```
remove the ENROLL_COMMAND_MAX_LEN and update docs to calculate strlen before allocation

srieteja commented on packages/atcommons/tests/test_enroll_namespace_utils.c:
```diff
@@ -0,0 +1,95 @@
+  if (ret != 0) {
+    printf("test failed\nexpected: %s\n*actual*: %s\n", en_expected_json, ns_json);
+  }
+
```
relpace printf's with atlogger

srieteja commented on packages/atauth/src/atactivate.c:
```diff
@@ -0,0 +1,404 @@
+  unsigned char *encrypted_self_encryption_key = malloc(aes256_encrypted_aes256_key_unsigned_char_size);
+
+  // intialize base64 encoded encrypted APKAM symmetric Key and encrypted default encryption private key
+  const size_t aes256_encrypted_rsa_2048_privkey_base64_len = atchops_base64_encoded_size(rsa_2048_privkey_base64_len);
```
A second aes-ctr is missing

xavierchanth replied:
Can you explain why this was resolved without making the changes

srieteja commented on packages/atauth/src/atactivate.c:
```diff
@@ -0,0 +1,404 @@
+  /*
+   * 6. Perform PKAM auth
+   */
+  printf("atisgm: %s\n", atsign);
```
remove debug log

srieteja commented on packages/atauth/src/atactivate.c:
```diff
@@ -0,0 +1,404 @@
+  /*
+   * 8. Delete CRAM secret from the server
+   */
+  atclient_atkey_free(&atkey);
```
create a new pointer

srieteja committed:
chore: address review comments - round 2

srieteja commented:
> See how I defined a `size` and `len` and sent the `len`
> 
> ```c
>   const size_t update_cmd_size =
>       strlen("update") + metadata_protocol_str_len + strlen(":") + atkey_str_len + strlen(" ") + value_encrypted_base64_len + strlen("\r\n") + 1;
>   if ((update_cmd = malloc(sizeof(char) * update_cmd_size)) == NULL) {
>     ret = 1;
>     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to allocate memory for update_cmd\n");
>     goto exit;
>   }
> 
>   snprintf(update_cmd, update_cmd_size, "update%s:%s %s\r\n", metadata_protocol_str, atkey_str, value_encrypted_base64);
>   const size_t update_cmd_len = update_cmd_size - 1;
> 
>   /*
>    * 7. Send update command
>    */
>   if ((ret = atclient_connection_send(&ctx->atserver_connection, (unsigned char *) update_cmd, update_cmd_len, recv, recv_size,
>                                       &recv_len)) != 0) {
>     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_connection_send: %d\n", ret);
>     goto exit;
>   }
> ```

I tried this approach for the cram command. It worked. Thanks, @JeremyTubongbanua

srieteja committed:
fix: atsign memory leak bug

srieteja committed:
fix: atsign memory leak bug

srieteja committed:
Merge remote-tracking branch 'origin/create_at_commons' into create_at_commons

# Conflicts:
#	packages/atauth/src/atactivate.c

xavierchanth commented on examples/desktop/at_talk/src/main.c:
```diff
@@ -15,6 +15,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
+#include <bits/getopt_core.h>
```
private header, see man getopt(3)

xavierchanth commented on packages/atauth/include/atauth/send_enroll_request.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef ATAUTH_SEND_ENROLL_REQUEST
+#define ATAUTH_SEND_ENROLL_REQUEST
+
+#include "../../../atclient/include/atclient/atclient.h"
+// #include <atclient/atclient.h>
```
Incorrect header

xavierchanth commented on packages/atauth/include/atauth/cjson.h:
```diff
@@ -0,0 +1,10 @@
+#ifndef ATCLIENT_CJSON_H
+#define ATCLIENT_CJSON_H
```
define should be atauth_cjson, but also atclient is a dependency of atauth, no need to duplicate this file, just `#include atclient/cjson.h`

xavierchanth commented on packages/atchops/include/atchops/hex.h:
```diff
@@ -0,0 +1,38 @@
+#ifndef UTILS_H
+#define UTILS_H
```
Macro guards for pragma once should follow the naming scheme of `<PACKAGENAME>_<FILENAME>` e.g. `ATCHOPS_HEX_H`, please check all other headers

xavierchanth replied:
This file is also missing `__cplusplus` extern macros, see sha.h for an example at the top and bottom of the file.

xavierchanth commented on packages/atchops/include/atchops/utf8.h:
```diff
@@ -0,0 +1,14 @@
+#ifndef UTF8_H
+#define UTF8_H
```
Same as comments for hex.h

xavierchanth commented on packages/atauth/CMakeLists.txt:
```diff
@@ -0,0 +1,146 @@
+# Configurable options
+option(ATAUTH_BUILD_TESTS "Build tests for atauth" OFF)
+
```
Since atauth is both a library and produces the at_activate executable, there needs to be an option which by default doesn't generate the executable.

xavierchanth commented on packages/atclient/include/atclient/atkeys_file.h:
```diff
@@ -1,8 +1,5 @@
-#ifdef __cplusplus
-extern "C" {
-#endif
+#ifndef ATCLIENT_atkeys_file_H
+#define ATCLIENT_atkeys_file_H
```
This should not have been changed

xavierchanth commented on packages/atclient/src/atclient.c:
```diff
@@ -391,6 +392,214 @@ exit: {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_string_utils_atsign_with_at: %d\n", ret);
+    goto exit;
+  }
+  char *atsign_without_at = malloc(sizeof(char) * strlen(atsign_with_at) + 1);
```
check for null

xavierchanth commented on packages/atclient/src/atclient.c:
```diff
@@ -391,6 +392,214 @@ exit: {
+  /*
+   * 10a. Build `cram:` noop_cmd
+   */
+  cram_cmd = malloc(sizeof(char) * ATCLIENT_CRAM_COMMAND_LEN + 1); // free later
```
check for null

xavierchanth commented on packages/atclient/src/atkeys_file.c:
```diff
@@ -145,12 +145,14 @@ int atclient_atkeys_file_from_string(atclient_atkeys_file *atkeys_file, const ch
-      atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "set_apkam_symmetric_key_str: %d\n", ret);
-      goto exit;
-    }
+  if (apkam_symmetric_key == NULL) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to read apkamSymmetricKey from JSON\n");
+    goto exit;
+  }
+  if ((ret = set_apkam_symmetric_key_str(atkeys_file, apkam_symmetric_key->valuestring,
+                                         strlen(apkam_symmetric_key->valuestring))) != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "set_apkam_symmetric_key_str: %d\n", ret);
+    goto exit;
```
The change is incorrect, not every keyfile will have apkamSymmetricKey set, but they are still valid files if they do not have the value.

xavierchanth commented on packages/atclient/src/connection.c:
```diff
@@ -402,11 +403,55 @@ int atclient_connection_send(atclient_connection *ctx, const unsigned char *src,
+    int attempt = 0;
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_DEBUG,
+                 "Partial write successful. Performing recursive write. Current ret: %d\n", ret);
+    // warning: potiential infinite loop
```
It's not though, ret will always decrease or return a negative number.

If it writes nothing, that's an error and it will return a negative code.
If it writes something, ret will decrease.
Therefore ret will always reach 0 or a negative number, which will exit the loop.

This is also missing a check for if ret == MBEDTLS_ERR_SSL_WANT_WRITE. Which is another scenario where ret is negative, and the loop should continue.

xavierchanth replied:
The correct pattern for implementing mbedtls_ssl_write should already be elsewhere in the code.

srieteja replied:
I could not find the existing impl for this. So, for now I have reverted this change as part of [9e57229](https://github.com/atsign-foundation/at_c/pull/423/commits/9e57229bbdbd7f1ec1a8fe0590500925bfa4e110). And since this is not causing any issues yet, we can implement this as part of a different ticket in the future.

xavierchanth commented on packages/atclient/src/pkam_auth_ex.c:
Empty file

xavierchanth commented on packages/atclient/src/request_options.c:
```diff
@@ -1200,6 +1202,119 @@ void atclient_get_public_key_request_options_unset_store_atkey_metadata(
+   * 1. Validate arguments
+   */
+  if (options == NULL) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_delete_request_options_free: Invalid arguments\n");
```
No need to log here, just return

xavierchanth commented on packages/atcommons/include/atcommons/at_expect.h:
```diff
@@ -0,0 +1,6 @@
+#ifndef ATCOMMONS_EXPECT_H
+#define ATCOMMONS_EXPECT_H
+
+int atcommons_string_expect(char *actual, char *expected);
+
+#endif //ATCOMMONS_EXPECT_H
```
This file should not be in the include directory. This is a function which is defined and only used in the tests. This header file should be moved to the tests folder and it should be added as an include manually from the tests cmake file. This will cause an unresolved symbol when building with the tests disabled

xavierchanth commented on packages/atcommons/include/atcommons/cjson.h:
```diff
@@ -0,0 +1,10 @@
+#ifndef ATCLIENT_CJSON_H
```
Same thing, no need to define in multiple places, just use the already available header file

xavierchanth commented on packages/atcommons/include/atcommons/enroll_command_builder.h:
```diff
@@ -0,0 +1,24 @@
+#ifndef ENROLL_VERB_BUILDER_H
+#define ENROLL_VERB_BUILDER_H
```
ATCOMMONS_

xavierchanth commented on packages/atcommons/include/atcommons/enroll_command_builder.h:
```diff
@@ -0,0 +1,24 @@
+#define ENROLL_VERB_BUILDER_H
+
+#include "enroll_operation.h"
+#include "enroll_params.h"
```
don't use relative headers, should be of the format: `#include "atcommons/*.h"`

xavierchanth commented on packages/atcommons/include/atcommons/enroll_namespace.h:
```diff
@@ -0,0 +1,61 @@
+typedef struct {
+  char *name;
+  char *access;
+} enroll_namespace_t;
```
atcommons_

xavierchanth commented on packages/atcommons/include/atcommons/enroll_namespace.h:
```diff
@@ -0,0 +1,61 @@
+#ifndef ENROLL_NAMESPACE_STRUCT_H
+#define ENROLL_NAMESPACE_STRUCT_H
```
ATCOMMONS_

xavierchanth commented on packages/atcommons/include/atcommons/enroll_namespace.h:
```diff
@@ -0,0 +1,61 @@
+typedef struct {
+  size_t length;
+  enroll_namespace_t *namespaces[];
+} enroll_namespace_list_t;
```
atcommons_

xavierchanth commented on packages/atcommons/include/atcommons/enroll_operation.h:
```diff
@@ -0,0 +1,25 @@
+ * @param e_op enroll operation as enum enroll_operation_t
+ * @return int 0 on success, non-zero on failure
+ */
+int enroll_operation_to_string(char **op_name, enroll_operation_t e_op);
```
atcommons_

xavierchanth commented on packages/atcommons/include/atcommons/enroll_operation.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef ENROLL_OPERATION_H
+#define ENROLL_OPERATION_H
```
ATCOMMONS_

xavierchanth commented on packages/atcommons/include/atcommons/enroll_operation.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef ENROLL_OPERATION_H
+#define ENROLL_OPERATION_H
+
+#define MAX_ENROLL_OPERATION_STRING_LEN 8
```
ATCOMMONS_

xavierchanth commented on packages/atcommons/include/atcommons/enroll_params.h:
```diff
@@ -0,0 +1,40 @@
+#ifndef ENROLL_PARAMS_H
+#define ENROLL_PARAMS_H
```
Last one I'm marking, ATCOMMONS_

xavierchanth commented on packages/atcommons/include/atcommons/enroll_params.h:
```diff
@@ -0,0 +1,40 @@
+#ifndef ENROLL_PARAMS_H
+#define ENROLL_PARAMS_H
+
+#include "enroll_namespace.h"
```
dont use relative imports

xavierchanth commented on packages/atcommons/src/enroll/enroll_namespace.c:
```diff
@@ -0,0 +1,81 @@
+    return -1;
+  }
+  // Create a new cJSON object
+  cJSON *json_obj = cJSON_CreateObject();
```
LEAK: this only gets freed when there is an error with cJSON_PrintUnformatted

xavierchanth commented on packages/atcommons/src/enroll/enroll_params.c:
```diff
@@ -0,0 +1,126 @@
+   * 2. Initialize
+   */
+  memset(ep, 0, sizeof(enroll_params_t));
+  ep->enrollment_id = NULL;
+  ep->app_name = NULL;
+  ep->device_name = NULL;
+  ep->otp = NULL;
+  ep->apkam_public_key = NULL;
+  ep->encrypted_default_encryption_private_key = NULL;
+  ep->encrypted_self_encryption_key = NULL;
+  ep->encrypted_apkam_symmetric_key = NULL;
+  ep->ns_list = NULL;
```
Redundant operations, memset'ing everything to 0 already makes everything NULL

xavierchanth commented on packages/atcommons/src/enroll/enroll_params.c:
```diff
@@ -0,0 +1,126 @@
+    return ret;
+  }
+
+  cJSON *json_object = cJSON_CreateObject();
```
LEAK: not freed

xavierchanth reviewed (changes requested):
See comments above.

When making large changesets please avoid making changes that aren't related to the main purpose for the PR. It's fine if it's to fix something that you need, but that wasn't always the case in this changeset.

srieteja committed:
feat: introduce auth_cli

srieteja committed:
Revert "feat: introduce auth_cli"

This reverts commit a3d043af994893570219a361a7d2b8daeb12ff42.
This reverts commit d646b46a4811fef088c63452ca0df6cd75ded9ca.
This reverts commit aff312e53e9457387ad92d648b21f96021703128.
This reverts commit e8c7b71a3aca356e1ba680c4bf8af00f960cbd8d.

srieteja committed:
fix: address review comments

srieteja committed:
chore: revert changes reg recursive write in atclient/connection.c

srieteja commented:
All of the review comments have been adressed as part of [5162fb8](https://github.com/atsign-foundation/at_c/pull/423/commits/5162fb8572af813f821080d2b3d1b06c27948c3b).
